packages/imt-rs/src/lib.rs

```diff
@@ -0,0 +1,320 @@
+use std::error::Error;
+
+use tiny_keccak::{Hasher, Keccak};
+
+pub struct IMT {
+    nodes: Vec<Vec<Option<IMTNode>>>,
+    zeroes: Vec<IMTNode>,
+    hash: IMTHashFunction,
+    depth: usize,
+    arity: usize,
+}
+
+struct IMTMerkleProof {
+    root: IMTNode,
+    leaf: IMTNode,
+    path_indices: Vec<usize>,
+    siblings: Vec<Vec<IMTNode>>,
+    leaf_index: usize,
+}
+
+
+impl IMT {
+    pub fn new(hash: IMTHashFunction, depth: usize, zero_value: IMTNode, arity: usize, leaves: Option<Vec<IMTNode>>) -> Result<IMT, &'static str> {
+
+        if leaves.is_some() && leaves.as_ref().unwrap().len() > arity.pow(depth as u32) {
+            return Err("The tree cannot contain more than arity^depth leaves");
+        }
+
+        let mut imt = IMT {
+            nodes: vec![vec![]; depth + 1],
+            zeroes: vec![],
+            hash,
+            depth,
+            arity,
+        };
+
+        let mut current_zero = zero_value;
+        for _ in 0..depth {
+            imt.zeroes.push(current_zero.clone());
+            current_zero = (imt.hash)(vec![current_zero; arity]);
+        }
+
+        if let Some(leaves) = leaves {
+            imt.nodes[0] = leaves.into_iter().map(Some).collect();
+
+            for level in 0..depth {
+                for index in 0..((imt.nodes[level].len() as f64 / arity as f64).ceil() as usize) {
+                    let position = index * arity;
+                    let children: Vec<_> = (0..arity)
+                        .map(|i| imt.nodes[level].get(position + i).cloned().flatten().unwrap_or_else(|| imt.zeroes[level].clone()))
+                        .collect();
+
+                    if let Some(next_level) = imt.nodes.get_mut(level + 1) {
+                        next_level.push(Some((imt.hash)(children)));
+                    }
+                }
+            }
+        }
+
+        Ok(imt)
+    }
+
+    pub fn root(&mut self) -> Option<String>  {
+        let nodes = self.nodes[self.depth].clone();
+        return nodes[0].clone();
+    }
+
+    pub fn depth(&mut self) -> usize {
+        return self.depth;
+    }
+
+    pub fn leaves(&mut self) -> Vec<Option<String>> {
+        return self.nodes[0].clone();
+    }
+
+    pub fn arity(&mut self) -> usize {
+        return self.arity;
+    }
+
+    pub fn insert(&mut self, leaf: IMTNode) -> Result<(), &'static str> {
```

Comment:
I think would be good to try to change insert function with insertion to vec, and then calling the `update` function to update/recalculate the tree